examples/websocket-relay/relay-app/src/lib/workflow.ts

```diff
@@ -61,27 +58,56 @@ export async function run(workflowId: WorkflowId) {
 
   // Send run command to server
   if (workflowId === "one") {
-    channel?.send(
-      JSON.stringify({
-        action: "run",
-        name: workflowId,
-        workflow: workflowOneJson,
-      })
-    );
+    const workflowOne = await workflowOnePromised;
+    homestar.runWorkflow(workflowOne, handleMessage);
   } else if (workflowId === "two") {
-    channel?.send(
-      JSON.stringify({
-        action: "run",
-        name: workflowId,
-        workflow: workflowTwoJson,
-      })
-    );
+    const workflowTwo = await workflowTwoPromised;
+    homestar.runWorkflow(workflowTwo, handleMessage);
   }
 
-  if (import.meta.env.VITE_EMULATION_MODE === "true") {
-    // Emulate with an echo server
-    emulate(workflowId, channel);
-  }
+  checkHealth();
+}
+
+/**
+ * Check health and fail workflow when the Homestar node does not respond in time.
+ */
+function checkHealth() {
+  const homestar = getStore(homestarStore);
+
+  let interval = setInterval(async () => {
+    const activeWorkflow = getStore(activeWorkflowStore);
+
+    if (activeWorkflow) {
+      if (activeWorkflow.step === activeWorkflow.tasks.length - 1) {
+        // Workflow completed
+        clearInterval(interval);
+      }
+
+      if (
+        activeWorkflow.failedPingCount >= import.meta.env.VITE_MAX_PING_RETRIES
+      ) {
+        // Fail the workflow
+        fail(activeWorkflow.id);
+        clearInterval(interval);
+      }
+
+      const health = (await homestar.health()).result;
+      if (health?.healthy) {
+        activeWorkflowStore.update((store) =>
+          store ? { ...store, failedPingCount: 0 } : null
+        );
+      } else {
+        activeWorkflowStore.update((store) =>
+          store
+            ? { ...store, failedPingCount: activeWorkflow.failedPingCount + 1 }
+            : null
+        );
+      }
+    } else {
+      // No workflow active
+      clearInterval(interval);
+    }
+  }, import.meta.env.VITE_PING_INTERVAL);
 }
```

Comment:
we can add this feature to the client itself.
you should actually subscribe to the client `error` events to handle unexpected behaviour

Reply:
Yeah! That would be great. 💯